aioh2/protocol.py

```diff
@@ -362,6 +365,23 @@ def _priority_step(self):
 
     # APIs
 
+    def set_handler(self, handler):
+        """
+        Connect with a coroutine, which is called when connection is made.
```

Comment:
This is not entirely true - the `handler` is actually called ***soon*** (if not immediately) after `set_handler` because it is fed to `asyncio.async`.

I believe you were trying to deal with a dirty shutdown of the server side here - however asyncio did this differently:

```python
import asyncio

async def cb(r, w):
    while True:
        line = await r.readline()
        if line:
            print(line)
        else:
            break

loop = asyncio.get_event_loop()
loop.run_until_complete(asyncio.start_server(cb, port=32123))

import socket
s = socket.socket()
s.connect(('0.0.0.0', 32123))
s.send(b'haha\nhehe\n')
s.close()

loop.run_until_complete(asyncio.sleep(1))
```

It closes the `Task` internally (and up to the user of the library), not brutally `cancel`s the `Task` (forced in the library). Your PR inspires the following: can we (or is it proper to) somehow raise an exception during all `await`ing calls to the `H2Protocol` when a disconnection is detected, where the exceptions will then by default lead to an exit of the pending user's `Task`s, also giving users the opportunity to do cleanups on such cases. However, this doesn't look like an easy change - might need to implement `Event` that raises exceptions and `Queue` that can be closed.